File: app/src/main/java/com/autotune/arabic/engine/PitchDetector.java
package com.autotune.arabic.engine;

/**
 * كاشف الطبقة الصوتية باستخدام خوارزمية YIN.
 */
public class PitchDetector {

    public static final double NO_PITCH = -1.0;

    private static final double YIN_THRESHOLD = 0.15;
    private static final double MIN_ENERGY    = 1e-6;

    private final int     sampleRate;
    private final int     bufferSize;
    private final double[] yin;
    private final float[]  linear;  // نسخة مرتبة زمنياً من ring buffer

    public PitchDetector(int sampleRate, int bufferSize) {
        this.sampleRate = sampleRate;
        this.bufferSize = bufferSize;
        this.yin    = new double[bufferSize / 2];
        this.linear = new float[bufferSize];
    }

    /**
     * @param ring     ring buffer مُعبَّأ دائرياً
     * @param writePos موضع الكتابة التالي (= أقدم عينة في الـ ring)
     */
    public double detect(float[] ring, int writePos) {
        // نسخ الـ ring buffer بالترتيب الزمني الصحيح (أقدم → أحدث)
        int tail = bufferSize - writePos;
        System.arraycopy(ring, writePos, linear, 0,    tail);
        System.arraycopy(ring, 0,        linear, tail, writePos);

        int half = bufferSize / 2;

        double energy = 0;
        for (int i = 0; i < bufferSize; i++) energy += (double) linear[i] * linear[i];
        if (energy / bufferSize < MIN_ENERGY) return NO_PITCH;

        // دالة الفرق
        for (int tau = 0; tau < half; tau++) {
            yin[tau] = 0;
            for (int j = 0; j < half; j++) {
                double d = linear[j] - linear[j + tau];
                yin[tau] += d * d;
            }
        }

        // المتوسط التراكمي المُعيَّر
        yin[0] = 1.0;
        double runSum = 0;
        for (int tau = 1; tau < half; tau++) {
            runSum += yin[tau];
            yin[tau] = runSum > 0 ? yin[tau] * tau / runSum : 1.0;
        }

        int tauMin = sampleRate / 1200;
        int tauMax = Math.min(half - 2, sampleRate / 80);

        int tauEst = -1;
        for (int tau = tauMin; tau <= tauMax; tau++) {
            if (yin[tau] < YIN_THRESHOLD) {
                while (tau + 1 <= tauMax && yin[tau + 1] < yin[tau]) tau++;
                tauEst = tau;
                break;
            }
        }

        if (tauEst == -1) return NO_PITCH;

        double betterTau;
        if (tauEst > 1 && tauEst < half - 1) {
            double s0 = yin[tauEst - 1];
            double s1 = yin[tauEst];
            double s2 = yin[tauEst + 1];
            double den = 2.0 * (2.0 * s1 - s0 - s2);
            betterTau = Math.abs(den) > 1e-12 ? tauEst + (s2 - s0) / den : tauEst;
        } else {
            betterTau = tauEst;
        }

        return sampleRate / betterTau;
    }
}

File: ios/AutoTuneArabic/Engine/PitchDetector.swift
import Foundation

final class PitchDetector {
    static let noPitch: Double = -1.0
    private static let yinThreshold: Double = 0.15
    private static let minEnergy: Double = 1e-6

    private let sampleRate: Int
    private let bufferSize: Int
    private var yin: [Double]
    private var linear: [Float]

    init(sampleRate: Int, bufferSize: Int) {
        self.sampleRate = sampleRate
        self.bufferSize = bufferSize
        self.yin = [Double](repeating: 0.0, count: bufferSize / 2)
        self.linear = [Float](repeating: 0.0, count: bufferSize)
    }

    func detect(ring: [Float], writePos: Int) -> Double {
        // Copy ring buffer to linear array in chronological order
        let tail = bufferSize - writePos
        linear.withUnsafeMutableBufferPointer { linBuf in
            ring.withUnsafeBufferPointer { ringBuf in
                // Copy from writePos to end
                for i in 0..<tail {
                    linBuf[i] = ringBuf[writePos + i]
                }
                // Copy from start to writePos
                for i in 0..<writePos {
                    linBuf[tail + i] = ringBuf[i]
                }
            }
        }

        let half = bufferSize / 2

        // Energy check
        var energy: Double = 0.0
        for i in 0..<bufferSize {
            let s = Double(linear[i])
            energy += s * s
        }
        if energy / Double(bufferSize) < PitchDetector.minEnergy {
            return PitchDetector.noPitch
        }

        // Difference function
        for tau in 0..<half {
            yin[tau] = 0.0
            for j in 0..<half {
                let d = Double(linear[j]) - Double(linear[j + tau])
                yin[tau] += d * d
            }
        }

        // Cumulative mean normalized difference function
        yin[0] = 1.0
        var runSum: Double = 0.0
        for tau in 1..<half {
            runSum += yin[tau]
            yin[tau] = runSum > 0.0 ? yin[tau] * Double(tau) / runSum : 1.0
        }

        // Find minimum below threshold
        let tauMin = sampleRate / 1200
        let tauMax = min(half - 2, sampleRate / 80)
        var tauEst = -1
        var tau = tauMin
        while tau <= tauMax {
            if yin[tau] < PitchDetector.yinThreshold {
                while tau + 1 <= tauMax && yin[tau + 1] < yin[tau] {
                    tau += 1
                }
                tauEst = tau
                break
            }
            tau += 1
        }

        if tauEst == -1 {
            return PitchDetector.noPitch
        }

        // Parabolic interpolation
        let betterTau: Double
        if tauEst > 1 && tauEst < half - 1 {
            let s0 = yin[tauEst - 1]
            let s1 = yin[tauEst]
            let s2 = yin[tauEst + 1]
            let den = 2.0 * (2.0 * s1 - s0 - s2)
            betterTau = abs(den) > 1e-12 ? Double(tauEst) + (s2 - s0) / den : Double(tauEst)
        } else {
            betterTau = Double(tauEst)
        }

        return Double(sampleRate) / betterTau
    }
}
